nanshe_workflow/data.py

```diff
@@ -526,3 +527,300 @@ def astype(self, dtype):
+
+            Note:
+                Use ``pop`` if the data needs to kept around outside the store
+                after removal.
+        """
+
+        if key not in self:
+            raise KeyError(key)
+
+        # Cancel any existing writing tasks.
+        try:
+            value = self._memstore.pop(key)
+        except KeyError:
+            pass
+        else:
+            with dask.distributed.client.temp_default_client(self._client):
+                with dask.set_options(get=self._client.get):
+                    self._client.cancel(value, force=True)
+
+        del self._diskstore[key]
+
+    def __iter__(self):
+        return iter(self._diskstore)
+
+    def __len__(self):
+        return len(self._diskstore)
+
+    def __contains__(self, key):
+        return (key in self._diskstore)
+
+    def _create_dataset(self, name, shape, dtype, chunks, **kwargs):
+        """
+            Deletes the key specified
+
+            Cancels any existing computations immediately and invalidates any
+            references to the data in Dask. Blocks to delete the data from
+            disk. However depending on how this is implemented in the store it
+            may be fast.
+
+            Args:
+                name (str): name/key for the dataset in the store
+                shape (tuple of ints): space required for the dataset
+                dtype (type): type required for the dataset
+                chunks (tuple of ints): size of each chunk
+
+            Keyword Args:
+                **kwargs: miscellaneous other arguments
+
+            Returns:
+                Dataset allocated in disk store
+
+            Note:
+                Just uses ``create_dataset`` method of store.
+                Can override our method through subclassing.
+        """
+
+        r = self._diskstore.create_dataset(
+            name, shape=shape, dtype=dtype, chunks=chunks, **kwargs
+        )
+        return r
+
+    __marker = object()
+
+    def pop(self, key, default=__marker):
+        """
+            Pops the value associated with the key
+
+            Loads the data associated with the specified key into
+            Distributed's memory. Then deletes the data from the disk store and
+            cancels any Futures related to the old value.
+
+            As a result, this can be very slow and is synchronous. Users can
+            preempt the process by persisting earlier if they know this will be
+            needed.
+
+            Args:
+                key (str): key to pop
+                default (Dask Array): value to return if key is missing
+
+            Returns:
+                Dask Array: Value that was in store
+
+            Note:
+                This should ensure that older references to this data are
+                still valid. However this may involve tricks to get rid of the
+                canceled Futures.
+
+                xref: https://github.com/dask/distributed/issues/1861
```

Comment:
xref: https://github.com/dask/distributed/issues/1861

Reply:
Admittedly with newer copies of Distributed am not seeing this as much of an issue. Distributed 1.21.4 seemed fine.